drv/lpc55-update-server/src/main.rs

```diff
@@ -385,6 +357,89 @@ impl idl::InOrderUpdateImpl for ServerImpl<'_> {
     }
 }
 
+impl ServerImpl<'_> {
+    // Locate and return the authoritative CFPA flash word number.
+    //
+    // There are two "official" copies of the CFPA, referred to as
+    // ping and pong. One of them will supercede the other, based on
+    // a monotonic version field at offset 4. We'll take the
+    // contents of whichever one is most recent, alter them, and
+    // then write them into the _third_ copy, called the scratch
+    // page.
+    //
+    // At reset, the boot ROM will inspect the scratch page, check
+    // invariants, and copy it to overwrite the older of the ping
+    // and pong pages if it approves.
+    //
+    // That means you can apply this operation several times before
+    // resetting without burning many monotonic versions, if you
+    // want to do that for some reason.
+    //
+    // The addresses of these pages are as follows (see Figure 13,
+    // "Protected Flash Region," in UM11126 rev 2.4, or the NXP
+    // flash layout spreadsheet):
+    //
+    // Page     Addr        16-byte word number
+    // Scratch  0x9_DE00    0x9DE0
+    // Ping     0x9_E000    0x9E00
+    // Pong     0x9_E200    0x9E20
+    fn cfpa_word_number(&mut self) -> Result<u32, UpdateError> {
+        // Read the two versions. We do this with smaller buffers so
+        // we don't need 2x 512B buffers to read the entire CFPAs.
+        let mut ping_header = [0u32; 4];
+        let mut pong_header = [0u32; 4];
+
+        indirect_flash_read(
+            &mut self.flash,
+            0x9E00,
+            core::slice::from_mut(&mut ping_header),
+        )?;
+        indirect_flash_read(
+            &mut self.flash,
+            0x9E20,
+            core::slice::from_mut(&mut pong_header),
+        )?;
+
+        // Work out where to read the authoritative contents from.
+        let val = if ping_header[1] >= pong_header[1] {
+            0x9E00
+        } else {
+            0x9E20
+        };
+
+        Ok(val)
+    }
+
+    /// Return the persistent and transient boot preferences
+    fn boot_preferences(
+        &mut self,
+    ) -> Result<(SlotId, Option<SlotId>), UpdateError> {
+        let cfpa_word_number = self.cfpa_word_number()?;
```

Comment:
I realized yesterday that this likely *not* the behavior we want. The factoring out of `self.cfpa_word_number()` only returns the correct version of cfpa page by looking at the `ping` and `pong` pages. Unfortunately, when we update the CFPA permanent boot selection we write to the `scratch` page. This only gets promoted to the higher version of `ping` or `pong` and becomes authoritative after a reboot. Therefore an operator, say, selects slot `B` as the boot preference, and then goes and reads the boot preference we will show them the previously adopted authoritative page, which may show `A` as the preference.

What we actually want to do is find the current authoritative page, and also see if there has been an update to the scratch page by comparing versions. If the scratch page is more recent it means an update was written but we haven't rebooted and we should show what was written as it will become authoritative on boot.

Interestingly I also discovered, at least on this branch, that bootleby does not seem to be respecting the selection of bank B, and will always boot bank A unless I erase A and only leave B. 

CC @cbiffle